### PR TITLE
Port to C

### PR DESCRIPTION
**Pull request by irandms**
If the following results sound right for what the .ino arduino main file have for their test packets, then the controller command/data C files should be done. You can access commands using the global variable "command", ala

`command.a`
`command.b`

same for data:

`packets.a.crc`
`packets.b.rpm`

test_results.txt:

> bin/controllerdata_test 
> PACKET A VALUES
> Command: 58
> Static var: 16
> Throttle: 45
> Brake_pedal: 5
> Foot switch: 0
> Forward switch: 0
> reverse: 0
> hall_a: 1
> hall_b: 0
> hall_c: 0
> battery_voltage: 59
> motor_temp: 0
> controller_temp: 14
> setting_dir: 0
> actual_dir: 0
> crc: 198
> PACKET B VALUES
> RPM: 49155
> Phase current: 1792
> crc: 21

### CODE REVIEW

File: arduino/controllerdata.c [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+    }
+}
+
+void _bytes_to_packet(struct generic_packet *pkt, uint8_t *bytes) {
+    for(int i = 0; i < PACKET_LENGTH; i++) {
+        pkt->bytes[i] = bytes[i];
+    }
+}
+
+void import_bytes_to_packets(_packets_struct *pkts, uint8_t *bytes) {
```
Comment (irandms) [inline arduino/controllerdata.c:38]: Main externally usable function; call on a buffer containing serial read data

File: arduino/controllerdata.h [hunk 2]
```diff
@@ -1,37 +1,72 @@
...
+    uint8_t     padding_9;
+    uint8_t     padding_10;
+    uint8_t     padding_11;
+    uint8_t     padding_12;
+    uint8_t     padding_13;
+    uint8_t     padding_14;
+    uint8_t     crc;
+};
+
+struct generic_packet {
```
Comment (irandms) [inline arduino/controllerdata.h:55]: generic packets are cast to type-specific packet struct upon identification (Cast via memcpy)

File: klsreader/protocol/controllerdata.py [hunk 3]
```diff
@@ -40,7 +40,7 @@ def _unpackPacketA(self, data):
         (self.throttle, self.brakePedal, self.brakeSwitch, self.footSwitch,
         self.forwardSwitch, self.reverse, self.hallA, self.hallB, self.hallC,
         self.batteryVoltage, self.motorTemp, self.controllerTemp, self.settingDir,
-        self.actualDir) = unpack('2x5B4?3B2?3x', data)
+        self.actualDir) = unpack('!2x5B4?3B2?3x', data)
```
Comment (irandms) [inline klsreader/protocol/controllerdata.py:43]: See if this breaks anything and if so reject the PR and I can fix it, or you can recommit over this.

File: klsreader/protocol/controllerdata.py [hunk 4]
```diff
@@ -40,7 +40,7 @@ def _unpackPacketA(self, data):
...
 
     def _unpackPacketB(self, data):
-        (self.rpm, self.phaseCurrent) = unpack('>4x2H11x', data)
+        (self.rpm, self.phaseCurrent) = unpack('!4x2H11x', data)
```
Comment (irandms) [inline klsreader/protocol/controllerdata.py:46]: Big endian is network order, changed both to network order for clarity

File: klsreader/protocol/controllerdata.py (hunk 3 above)
Comment (hunter-young) [reply klsreader/protocol/controllerdata.py:43]: Doesn't seem to break anything, and I don't see what future harm could result from it. I'll implement it on the next pull request.

File: klsreader/protocol/controllerdata.py (hunk 4 above)
Comment (hunter-young) [reply klsreader/protocol/controllerdata.py:46]: That's fine.

**Comment by hunter-young** [comment]
This is great! The only issue I see is that the endianness for packet.b is wrong. The RPM and Phase Current result values are little endian - they should be big endian. The correct values for RPM and Phase Current for the data packets are 960 and 7 respectively.

**Pull request closed without merging**